protocol-correctness/lib/language/map-utils.k

```diff
@@ -0,0 +1,165 @@
+require "protocol-correctness/lib/language/type-reflection.k"
+
+module MAP-UTILS-SYNTAX
+  imports MAP
+  imports TYPE-REFLECTION-SYNTAX
+
+  syntax KItem ::= concretizeMapKey(KItem, Map)
+
+  syntax KItem ::= nullableMapLookup(key:KItem, Map, ReflectionType)
+  syntax KItem ::= nullableMapSet(key:KItem, value:KItem, Map, ReflectionType)
+endmodule
+
+module MAP-UTILS
+  imports MAP-REFLECTION
+  imports MAP-UTILS-BASIC
+  imports MAP-UTILS-LANGUAGE
+endmodule
+
+module MAP-UTILS-SYNTAX-PRIVATE
+  imports MAP
+
+  syntax KItem ::= mapLookup(KItem, Map)
+  syntax KItem ::= mapDelete(key:KItem, Map)
+  syntax KItem ::= mapSet(key:KItem, value:KItem, Map)
+endmodule
+
+module MAP-UTILS-BASIC
+  imports BOOL
+  imports K-EQUAL-SYNTAX
+  imports MAP
+  imports MAP-UTILS-SYNTAX
+  imports MAP-UTILS-SYNTAX-PRIVATE
+  imports MAYBE
+  imports TYPE-REFLECTION-SYNTAX
+
+  rule concretizeMapKey(Key:KItem, M:Map) => splitMap(Key, M, ?_Value, ?_Remainder)
+    requires Key in_keys(M)
+    [label(concretizeMapKey)]
+  rule concretizeMapKey(Key:KItem, M:Map) => .K
+    requires notBool Key in_keys(M)
+    [label(concretizeKeyNotInMap)]
+
+
+  syntax KItem ::= "splitMapStuck" | "endSplitMap"
+  rule endSplitMap => .K
+
+
+  syntax KItem ::= splitMap(key:KItem, toSplit:Map, value:KItem, remainder:Map)
+      [function, functional]
+  rule splitMap(Key:KItem, M:Map, _Value:KItem, _Remainder:Map) => splitMapStuck
+    requires notBool Key in_keys(M)
+  rule splitMap(Key:KItem, (Key |-> _:KItem) M:Map, _Value:KItem, _Remainder:Map)
+      => endSplitMap
+    ensures notBool Key in_keys(M)
+    [simplification(30)]
+  rule splitMap(Key:KItem, (K1:KItem |-> _SomeValue:KItem) M:Map, Value:KItem, Remainder:Map)
+      => splitMap(Key, M, Value, Remainder)
+    requires notBool (Key ==K K1) andBool Key in_keys(M)
+    ensures notBool K1 in_keys(M)
+    [simplification(30)]
+  rule splitMap(Key:KItem, M:Map, Value:KItem, Remainder:Map)
+      => endSplitMap
+    requires Key in_keys(M)
+    ensures M ==K (Key |-> Value Remainder)
+    [simplification(50)]
+
+  // Do the map lookups in a way that makes the Haskell backend happy.
+
+  rule mapLookup(K:KItem, M:Map)
+      =>  concretizeMapKey(K, M) ~> #mapLookup(K, M)
+    requires K in_keys(M)
+    [label(mapLookupInKeys)]
+  rule mapLookup(K:KItem, M:Map)
+      =>  nothing
+    requires notBool K in_keys(M)
+    [label(mapLookupNotInKeys)]
+
+  syntax KItem ::= #mapLookup(KItem, Map)
+  rule #mapLookup(K:KItem, ((K |-> V:KItem) _:Map) #as M:Map) => just(V)
+    ensures K in_keys(M)
+  rule #mapLookup(K:KItem, M:Map) => nothing
+    ensures notBool (K in_keys(M))
+    [owise]
+
+  rule mapDelete(K:KItem, M:Map) => #mapDelete(K, M)
+    requires K in_keys(M)
+  rule mapDelete(K:KItem, M:Map) => M
+    requires notBool (K in_keys(M))
+
+  syntax KItem ::= #mapDelete(key:KItem, Map)
+  rule #mapDelete(K:KItem, (K |-> _:KItem) M:Map) => M
+  rule #mapDelete(_:KItem, M:Map) => M
```

Comment:
What does `ensures false` mean here?

Reply:
If this rule applies, it will remove the symbolic execution branch on which it was applied.

When applying `rule #mapDelete(K:KItem, (K |-> _:KItem) M:Map) => M` during symbolic execution, the Haskell backend will leave a remainder which will be stuck. The owise rule adds `false`/`bottom` to the path predicate, transforming the entire configuration into `bottom`

This could probably be solved by making the key concrete, as you mention above, but I think that I wrote the `concretizeMapKey` thing after implementing the set/lookup/delete operations for map, and I replaced it only where it was needed.

I'll try to get the full code in a stable state (right now I'm trying to prove a lemma), then I'll try to use `concretizeMapKey` here to see if anything changes. If everything is fine, I'll also replace it in this PR.

Reply:
got it